File: 백준/Silver/1149. RGB거리/RGB거리.java
/*
 * f(r, n) = min{f(g, n-1), f(b, n-1)} + nth house's cost of r
 * same for green, blue
 */

import java.io.*;
import java.util.*;

public class Main {
  public static void main (String[] args) {
  try {
    BufferedReader in = new BufferedReader(new InputStreamReader(System.in));

    // Initialize Houses
    int n = Integer.parseInt(in.readLine());
    // for include 0 and n+1 for easy algorithm
    House[] houses = new House[n+1]; 
    houses[0] = new House();
    for (int i = 1; i <= n; i++) {
      houses[i] = new House(in.readLine());
    }

    // Fill the table using the recurrence
    int[][] minCost = new int[n+1][3]; // for convenience of algorithm
    for (int i = 1; i <= n; i++) for (int c = 0; c < 3; c++)  {
      // Get current house cost
      int currCost = houses[i].getCost(Color.idxToColor(c));
      // Get minimum available cost of past houses
      int pastCost = -1;
      for (int k = 0; k < 3; k++) {
        if (k != c) {
          pastCost = (pastCost == -1 || pastCost > minCost[i-1][k])
            ? minCost[i-1][k] : pastCost;
        }
      }
      // Fill the table
      minCost[i][c] = currCost + pastCost;
    }

    System.out.println(Math.min(Math.min(minCost[n][0], minCost[n][1]), minCost[n][2]));

    BufferedWriter out = new BufferedWriter(new OutputStreamWriter(System.out));
  }
  catch (IOException e) {
    e.printStackTrace();
  }
  }
}

enum Color {
  R, G, B, NULL;

  public int getIdx() {
    switch (this) {
    case R: { return 0; }
    case G: { return 1; }
    case B: { return 2; }
    default: { return -1; }
    }
  }
  public static Color idxToColor(int i) {
    switch (i) {
    case 0 : { return R; }
    case 1 : { return G; }
    case 2 : { return B; }
    default: { return NULL; }
    }
  }
}

class House {
  private int RCost;
  private int GCost;
  private int BCost;
  private Color color = Color.NULL;

  public House() {
    RCost = GCost = BCost = 0;
  }
  public House(int rc, int gc, int bc) {
    RCost = rc; GCost = gc; BCost = bc;
  }

  public House(String costString) {
    int[] costs = Arrays.stream(
      costString.split(" "))
      .mapToInt(s -> Integer.parseInt(s))
      .toArray();

    RCost = costs[0]; GCost = costs[1]; BCost = costs[2];
  }

  public int getCost(Color c) {
    switch (c) {
    case R: {
      return RCost;
    }
    case G: {
      return GCost;
    }
    case B: {
      return BCost;
    }
    default: {
      return 0;
    }
    }
  }

  public Color getColor() {
    return color;
  }

  public void setColor(Color c) {
    color = c;
  }
}

File: 백준/Silver/1149. RGB거리/RGB거리.py
import sys
from enum import Enum

class Color(Enum):
    NONE = -1
    RED = 0
    GREEN = 1
    BLUE = 2

class House:
    def __init__(self, rp: int, gp: int, bp: int):
        self.red_price: int = rp
        self.green_price: int = gp
        self.blue_price: int = bp
    def getPrice(self, c: int) -> int:
        if c == Color.RED.value:
            return self.red_price
        elif c == Color.GREEN.value:
            return self.green_price
        elif c == Color.BLUE.value:
            return self.blue_price
        else:
            raise Exception

# initialize
n: int = int(sys.stdin.readline())
houses: list[House] = []
for i in range(n):
    houses.append(
        (lambda l : House(int(l[0]), int(l[1]), int(l[2]))) \
            (sys.stdin.readline().split(" "))
    )

# DP substructure
min_prices: list[list[int]] = [[] for _ in range(n)]
# for house 1
min_prices[0] = [houses[0].red_price, houses[0].green_price, houses[0].blue_price,]
# else
for i in range(1, n):
    for c in range(3):
            # v(i, c) = min(v(i-1, colors except c)) + price(i, c)
        min_prices[i].append(
            min(
                map(
                    lambda j : min_prices[i-1][j],
                    filter(
                        lambda x : x != c,
                        range(3)
                    )
                )
            )
            +
            houses[i].getPrice(c)
        )

print(min(min_prices[n-1]))